### DISCUSSION TITLE
Support non-index dataframe in heatmap

### DISCUSSION THREAD

**Discussion by stevenlis** [mwaskom/seaborn#3509] [Q&A]
Since Seaborn is planning to add support for dataframes other than pandas (https://github.com/mwaskom/seaborn/pull/3369), I'd like to point out an issue with heatmap.

Currently, Seaborn's [heatmap](https://seaborn.pydata.org/generated/seaborn.heatmap.html) requires a Pandas dataframe with row labels or an index. However, certain dataframes like Polars [do not have an index by design](https://pola-rs.github.io/polars/user-guide/migration/pandas/#selecting-data). It would be beneficial if the heatmap could provide an API that allows inputting a non-index dataframe.

```python
import pandas as pd
import polars as pl
import seaborn as sns
import matplotlib.pyplot as plt

data = {
    'A': [1, 2, 3],
    'B': [2, 3, 4],
    'C': [1, 3, 5],
    'Index': ['I', 'II', 'III']
}
```
```python
sns.heatmap(pd.DataFrame(data).set_index('Index'))
```
![download](https://github.com/mwaskom/seaborn/assets/33796896/242cdcaf-5cc8-4399-bed4-c830854d44bd)

For polars, the current workaround:
- set ticklabels in matplotlib manually

```python
df = pl.DataFrame(data)
fig, ax = plt.subplots()
sns.heatmap(df.drop('Index'), ax=ax)
ax.set_yticklabels(df.get_column('Index'))
```
- convert to pandas dataframe 
```python
sns.heatmap(pl.DataFrame(data).to_pandas().set_index('Index'))
```

**Comment by mwaskom** [comment]
I don't understand the request here. `heatmap` uses an index when it is available, but an index is not required, e.g. you can also pass a numpy array.

**Reply by stevenlis** [reply]
I'm not saying it doesn't work, but when you pivot a DataFrame in Polars, your index column will become just another column instead of an index object like in Pandas. This requires additional manipulation if you want to pass it to a seaborn heatmap, as shown above. I understand if this is not what Seaborn aims to cover.
